pkg/gc/gc_state_manager.go

```diff
@@ -543,6 +611,155 @@ func (m *GCStateManager) GetGCState(keyspaceID uint32) (GCState, error) {
+// This method simulates the behavior of the service safe points in old versions, by internally using txn safe points
+// and GC barriers. The behaviors are mapped as follows:
+//
+//   - The service safe point with service ID "gc_worker" is mapped to the txn safe point.
+//   - The service safe point with other service IDs are mapped to GC barriers with barrier IDs equal to the given
+//     service IDs.
+//
+// Note that the behavior of the service safe point of "gc_worker" is NOT perfectly the same as before: it can no longer
+// be advanced over other service safe points, but will be blocked by the minimal one; and if the cluster was running
+// with TiDB node that haven't migrated to the new GC APIs, it can also be blocked by the *TiDB min start ts* written by
+// those TiDB nodes.
+//
+// Therefore, the method's behavior is as follows:
+//
+//  1. If the given serviceID is "gc_worker", it internally calls AdvanceTxnSafePoint.
+//     - If the advancing result is the same as newServiceSafePoint, it's the case that the updated service safe
+//     point of "gc_worker" is exactly the minimal one. Returns a simulated service safe point with the service ID
+//     equals to "gc_worker".
+//     - Otherwise, it's the case that the service safe point of "gc_worker" is successfully updated, but it's not
+//     the minimal service safe point. Returns a simulated service safe point whose serviceID starts with
+//     "__pseudo_service:" to simulate the minimal service safe point. It may actually be either a GC barrier or
+//     a *TiDB min start ts*.
+//  2. If the given serviceID is anything else, it internally calls SetGCBarrier or DeleteGCBarrier, depending on
+//     whether the `ttl` is positive or not. As the txn safe point is always less or equal to any GC barriers, we
+//     simulate the case that the service safe point of "gc_worker" is the minimal one, and return a service safe point
+//     with the service ID equals to "gc_worker".
+//
+// This function only works on the NullKeyspace.
+func (m *GCStateManager) CompatibleUpdateServiceGCSafePoint(serviceID string, newServiceSafePoint uint64, ttl int64, now time.Time) (minServiceSafePoint *endpoint.ServiceSafePoint, updated bool, err error) {
+	keyspaceID := constant.NullKeyspaceID
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// TODO: After implementing the global GC barrier, redirect the invocation on "native_br" to `SetGlobalGCBarrier`.
+	if serviceID == keypath.GCWorkerServiceSafePointID {
+		if ttl != math.MaxInt64 {
+			return nil, false, errors.New("TTL of gc_worker's service safe point must be infinity")
+		}
+
+		res, err := m.advanceTxnSafePointImpl(keyspaceID, newServiceSafePoint, now)
+		if err != nil {
+			return nil, false, err
+		}
+		if res.NewTxnSafePoint != newServiceSafePoint {
+			minServiceSafePoint = &endpoint.ServiceSafePoint{
+				ServiceID: "__pseudo_service:" + res.BlockerDescription,
```

Comment:
Is it possible to unify the format of ServiceID? `__pseudo_service: TiDBMinStartTS { Key: %+q, MinStartTS: %d }` looks wired.

Reply:
Do you have any suggestion?